### PR TITLE
Path completion

### PR DESCRIPTION
**Pull request by sunjon**
Recursive path completion, RFC. Lots of rough edges, tried to avoid  dependencies for path and scandir functions.

### CODE REVIEW

File: lua/complementree/filters.lua [hunk 1]
```diff
@@ -32,4 +32,10 @@ M.strict_prefix = mk_filter(function(_, v, prefix)
   return vim.startswith(w, prefix) and #w ~= #prefix
 end)
 
+M.substr = mk_filter(function(_, v, prefix)
+  -- print(":: " .. prefix)
+  local w = utils.cword(v)
+  return w:find(prefix, 1, true) ~= nil
```
Comment (vigoux) [inline lua/complementree/filters.lua]: You can remove the `~= nil` here I think ?

File: lua/complementree/utils.lua [hunk 2]
```diff
@@ -13,4 +14,58 @@ function M.cword(complete_item)
...
+local function get_os_file_separator()
+  local os = string.lower(jit.os)
+  if os == "linux" or os == "osx" or os == "bsd" then
+    return "/"
+  else
+    return "\\"
+  end
+end
+
+local os_sep = get_os_file_separator()
```
Comment (vigoux) [inline lua/complementree/utils.lua]: As you do that only once, maybe avoid the function call and set `os_sep` correctly directly ?

File: lua/complementree/sources.lua [hunk 3]
```diff
@@ -162,6 +162,58 @@ local function lsp_completedone(completed_item)
   end
 end
 
+M.filepath = function(_, _, _, _)
+  local MAX_FILEPATH_DEPTH = 8 -- GET RID OF THESE
+  local MAKE_RELATIVE = true
```
Comment (vigoux) [inline lua/complementree/sources.lua]: Maybe turn the `filepath` function into a function that returns a source, and use the parameters of the outer function to configure the inner function ?

**Review by vigoux — COMMENTED** [review]
It looks definitely great, thank you very much for that !

**Comment by sunjon** [comment]
I'm testing it with a setup like this:
```lua
local fp_matcher = sources.create_filepath_matches {
  max_depth = 3,
  ignore_hidden = false,
  root_dirs = { "/home/sunjon" },
  match_patterns = { "%.png$", "%.jpg$" }
}
local path_completion = combine.pipeline(fp_matcher, filter.substr, sorter.alphabetic)
```
if `max_depth` key isn't specified, it defaults to `8`.
if `ignore_hidden` isn't specified, it defaults to `true`.
if `root_dirs` key isn't specified, it looks for LSP workspace directories, and then fallsback to CWD.
if `match_patterns` isn't specified, it defaults to matching everything.

using `source.filepath` returns a default filepath matcher.

File: lua/complementree/sources.lua [hunk 4]
```diff
@@ -162,17 +165,79 @@ local function lsp_completedone(completed_item)
   end
 end
 
+M.create_filepath_source = function(opts)
```
Comment (vigoux) [inline lua/complementree/sources.lua]: I think that we should keep `filepath_matches` for consistency here.

File: lua/complementree/sources.lua [hunk 5]
```diff
@@ -162,17 +165,79 @@ local function lsp_completedone(completed_item)
...
+  opts = opts or {}
+  local config = {
+    max_depth = opts.max_depth or 8,
+    ignore_hidden = opts.ignore_hidden or true,
+    relative_paths = opts.relative_paths or true,
+    root_dirs = opts.root_dirs,
+    match_patterns = opts.match_patterns or {}
+  }
+
+  local src = function(_, _, _, _)
```
Comment (vigoux) [inline lua/complementree/sources.lua]: Why not `return function...` here ?

Comment (vigoux) [reply lua/complementree/sources.lua]: And I think that this can be cached too ?

File: lua/complementree/sources.lua [hunk 6]
```diff
@@ -162,17 +165,80 @@ local function lsp_completedone(completed_item)
...
+        empty = 1,
+        user_data = { source = "filepath" },
+      }
+    end
+
+    return matches
+  end)
+end
+
+M.filepath = M.create_filepath_matches()
```
Comment (vigoux) [inline lua/complementree/sources.lua]: This should go into the `defaults.lua` and be of the form `M.filepath = pipeline(..., substr)` and such.

File: lua/complementree/sources.lua [hunk 7]
```diff
@@ -162,17 +165,80 @@ local function lsp_completedone(completed_item)
   end
 end
 
+M.create_filepath_matches = function(opts)
```
Comment (vigoux) [inline lua/complementree/sources.lua]: I would really like this like so:
```suggestion
M.filepath_matches = function(opts)
```

File: lua/complementree/defaults.lua [hunk 8]
```diff
@@ -21,4 +21,6 @@ M.luasnip = comb.pipeline(sources.luasnip_matches, filters.prefix, comp.alphabet
 
 M.lsp = comb.pipeline(sources.lsp_matches, filters.prefix, comp.alphabetic)
 
+M.filepath = comb.pipeline(sources.filepath_matches, filters.substr, comp.alphabetic)
```
Comment (vigoux) [inline lua/complementree/defaults.lua]:
```suggestion
M.filepath = comb.pipeline(sources.filepath_matches(), filters.substr, comp.alphabetic)
```

**Review by vigoux — CHANGES_REQUESTED** [review]

**Review by vigoux — COMMENTED** [review]
Could you add some tests for the substr filter ?

**Comment by sunjon** [comment]
Closing in favor of https://github.com/vigoux/complementree.nvim/pull/4

**Pull request closed without merging**